### DISCUSSION TITLE
Preface as an Ocaml base lib

### DISCUSSION THREAD

**Discussion by ildar** [xvw/preface#187] [General]
Ocaml is one of the rare programming languages having alternative core base std libraries.
I mean that a programmer has a choice of using Stdlib (which is the standard for Ocaml) OR using (Janestreet's) Base library (with Core as the superset). This alternative doesn't look marginal (as it could be) but seems to be quite widely used and supported.
I was greatly surprised when I got to know the Preface which looks like the second alternative.
But I'm not entirely sure that it is. So:
1. Is Preface complete enough to allow avoiding the Stdlib?
2. What are the possible conflicts when using Preface together with Stdlib?
3. Based on (2) what restrictions one can face if he/she decides using Preface as the base of a project?
I am sorry that some (or many) things above may be false. I've just touched the Preface and started reading the docs.

**Comment by xvw** [comment]
Hello, sorry for the delay (I was not familiar with Github's discussion). 
First of all, thanks a lot for your interest!

From my point of view, Preface is mostly "a kind of tool that deal with interfaces", even there is a `Preface.Stdlib` which was a design mistake. Currently we are thinking about Postface, a kind of Stdlib alternative that will fit with the same idea of Container or Base and relaying on Preface for usual abstraction. So to give a precise answer :

1. Nop, even the abstraction coverring is (imho) large, concretization are weak (this is why we want to design Postface)
2. opening Preface on top of a module is anoying because it shadows some modules so without making Preface opened on toplevel there is no issue
3. I know only two projects that uses Preface
    - https://github.com/xhtmlboi/yocaml that was created "in order to test the ergonomics of Preface"
    - https://github.com/xvw/muhokama mostly because it was built by the same people behind Preface

I hope that it answer to your questions and thanks for your interest!

**Comment by ildar** [comment]
I'm actively using Base in my pet project and like it peculiarities.
I also found an interesting project "Relude" which is an alternative for
Haskell's Prelude.
I think the Prelude's problems discussed in [this talk](
https://www.youtube.com/watch?v=qwAmiJ5M_zM) may be relevant for
Prelude/Postlude.
Thanks!
